### PR TITLE
sync-platform: Configure the platform / emulator to match circuits

### PR DESCRIPTION
**Pull request by naure**
Over time we moved from a first version (https://github.com/scroll-tech/ceno/issues/215) which looked end-to-end like this:

```
Static memory layout 
            `-->  Program compilation
            `-->  Trace generation
            `-->  Static circuits
```

to this

```
Existing third-party programs
            `-->  Auto-detected memory layout
                        `-->  Trace generation
                        `-->  Configurable circuits
```

This worked by letting the trace generation be more powerful and more permissive than the circuits (larger memory ranges, etc).

This PR configures the trace generation to tightly match the circuit behaviors as of today.

### CODE REVIEW

**Comment by matthiasgoergens** [comment]
Could you please expand the description with a short note on what aspects you are syncing up, please?

File: ceno_emul/src/platform.rs [hunk 1]
```diff
@@ -9,33 +9,42 @@ use crate::addr::{Addr, RegIdx};
 /// - codes of environment calls.
 #[derive(Clone, Debug)]
 pub struct Platform {
-    pub rom: Range<Addr>,
-    pub ram: Range<Addr>,
+    pub prog_code: Range<Addr>,
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:12]: We don't need `prog_code` in the platform.  There's no restrictions we need to enforce, apart from what's in the ELF, and we enforce them during decoding.

That's all thanks to the recent move to a Harvard Architecture.

File: ceno_emul/src/platform.rs [hunk 2]
```diff
@@ -9,33 +9,42 @@ use crate::addr::{Addr, RegIdx};
...
+    pub prog_data: Option<HashSet<Addr>>,
+    pub stack: Range<Addr>,
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:14]: I like the move to using a Range here!

File: ceno_emul/src/platform.rs [hunk 3]
```diff
@@ -9,33 +9,42 @@ use crate::addr::{Addr, RegIdx};
 /// - codes of environment calls.
 #[derive(Clone, Debug)]
 pub struct Platform {
-    pub rom: Range<Addr>,
-    pub ram: Range<Addr>,
+    pub prog_code: Range<Addr>,
+    pub prog_data: Option<HashSet<Addr>>,
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:13]: Consider changing to `BTreeSet` which has `pub const fn new() -> BTreeSet<T>` to make `const` objects.

File: ceno_emul/src/platform.rs [hunk 4]
```diff
@@ -9,33 +9,42 @@ use crate::addr::{Addr, RegIdx};
...
-    pub stack_top: Addr,
     /// If true, ecall instructions are no-op instead of trap. Testing only.
     pub unsafe_ecall_nop: bool,
 }
 
 pub const CENO_PLATFORM: Platform = Platform {
-    rom: 0x2000_0000..0x3000_0000,
-    ram: 0x8000_0000..0xFFFF_0000,
+    prog_code: 0x2000_0000..0x3000_0000,
+    prog_data: None, // This is an `Option` to allow `const` here.
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:24]: Move the comment to the type definition for Platform, please.  (Or change to `BTreeSet` and avoid the problem.)

File: ceno_emul/src/platform.rs [hunk 5]
```diff
@@ -9,33 +9,42 @@ use crate::addr::{Addr, RegIdx};
...
-    stack_top: 0xC0000000,
     unsafe_ecall_nop: false,
 };
 
 impl Platform {
     // Virtual memory layout.
 
-    pub fn is_rom(&self, addr: Addr) -> bool {
-        self.rom.contains(&addr)
+    pub fn is_prog_code(&self, addr: Addr) -> bool {
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:35]: We don't need this at all in the platform.  See the comment about Harvard architecture and decoding.

File: ceno_emul/src/platform.rs [hunk 6]
```diff
@@ -60,21 +69,21 @@ impl Platform {
...
     // Permissions.
 
     pub fn can_read(&self, addr: Addr) -> bool {
-        self.is_rom(addr) || self.is_ram(addr) || self.is_pub_io(addr) || self.is_hints(addr)
+        self.can_write(addr)
     }
 
     pub fn can_write(&self, addr: Addr) -> bool {
-        self.is_ram(addr)
+        self.is_ram(addr) || self.is_pub_io(addr) || self.is_hints(addr)
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:82]: `hints` (and pub io) are supposed be read-only.

File: ceno_zkvm/src/e2e.rs [hunk 7]
```diff
@@ -159,33 +160,79 @@ fn emulate_program(
...
+            stack: 0x0020_0400..0x0020_0400,
+            unsafe_ecall_nop: true,
+            ..CENO_PLATFORM
+        },
+    };
+
+    let prog_data = program.image.keys().copied().collect::<HashSet<Addr>>();
+    let stack = preset.stack.end - stack_size..preset.stack.end;
+    let heap = {
+        // Detect heap as starting after program data.
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/e2e.rs:189]: Btw, for ceno-compiled code ELFs, we have the `_sheap` and `_stack_start` symbols, which we can use read where heap and stack actually are.

We should also introduce symbols for the end of these, like we do for hints in https://github.com/scroll-tech/ceno/pull/631/files#diff-4994466418e9592df592e3378c56430a43d10451155f5f415da95ebe1a8699a7

So no need to come up with questionable heuristics for these values here.

File: ceno_zkvm/src/e2e.rs [hunk 8]
```diff
@@ -159,33 +160,79 @@ fn emulate_program(
...
+        let heap_start = program.image.keys().max().unwrap() + WORD_SIZE as u32;
+        let heap = heap_start..heap_start + heap_size;
+        // Pad the total size to the next power of two.
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/e2e.rs:192]: Please compare the discussion around https://github.com/scroll-tech/ceno/pull/716

Let's pick one policy for whether we should round up or down or assert, and decide one ideally one place where we do that.

**Review by matthiasgoergens — COMMENTED** [review]
Overall a nice consolidation and simplification!

Despite the title of the PR, I didn't see all that much new synchronisation between the platform and emulator here?  Perhaps I'm missing something.  Could you please elaborate in the PR description what new aspects we are matching that we didn't match before?  Thanks!

(I do mostly like the changes here in any case, even if I don't quite understand the title.)

**Comment by matthiasgoergens** [comment]
Please mention that we are doing this partially for SP1 compatibility in the PR description.

File: ceno_emul/src/platform.rs (hunk 1 above)
Comment (matthiasgoergens) [reply ceno_emul/src/platform.rs:12]: Hmm, all the comments here for this file seem to have moved down by one line, compared to where they should be.

File: ceno_emul/src/platform.rs [hunk 9]
```diff
@@ -60,21 +69,21 @@ impl Platform {
...
     pub const fn pc_base(&self) -> Addr {
-        self.rom.start
+        self.prog_code.start
     }
 
     // Permissions.
 
     pub fn can_read(&self, addr: Addr) -> bool {
-        self.is_rom(addr) || self.is_ram(addr) || self.is_pub_io(addr) || self.is_hints(addr)
+        self.can_write(addr)
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:78]: This completely precludes read-only memory?

File: ceno_emul/src/platform.rs [hunk 10]
```diff
@@ -60,21 +69,21 @@ impl Platform {
...
     }
 
     pub fn can_write(&self, addr: Addr) -> bool {
-        self.is_ram(addr)
+        self.is_ram(addr) || self.is_pub_io(addr) || self.is_hints(addr)
     }
 
     pub fn can_execute(&self, addr: Addr) -> bool {
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:85]: We don't need this anymore.

**Review by matthiasgoergens — CHANGES_REQUESTED** [review]
Please see comments.

File: ceno_emul/src/platform.rs (hunk 5 above)
Comment (naure) [reply ceno_emul/src/platform.rs]: Reverted.

File: ceno_emul/src/platform.rs (hunk 3 above)
Comment (naure) [reply ceno_emul/src/platform.rs]: Done.

File: ceno_emul/src/platform.rs (hunk 6 above)
Comment (naure) [reply ceno_emul/src/platform.rs:79]: They are not read-only so this matches correctly.

File: ceno_emul/src/platform.rs (hunk 9 above)
Comment (naure) [reply ceno_emul/src/platform.rs:75]: Yes, there are no read-only memory anywhere.

File: ceno_emul/src/platform.rs (hunk 10 above)
Comment (naure) [reply ceno_emul/src/platform.rs:82]: If you mean to merge `can_execute` into `fetch`, it should be removed in the PR that does that.

File: ceno_zkvm/src/e2e.rs (hunk 7 above)
Comment (naure) [reply ceno_zkvm/src/e2e.rs:189]: The "questionable heuristic" is the normal behavior of the linker.

I did consider switching to a symbol-based solution potentially that but that’s not the scope here.

**Review by lispc — APPROVED** [review]

File: ceno_emul/src/platform.rs (hunk 10 above)
Comment (matthiasgoergens) [reply ceno_emul/src/platform.rs:86]: Done.

File: ceno_emul/src/platform.rs (hunk 6 above)
Comment (matthiasgoergens) [reply ceno_emul/src/platform.rs:83]: OK.  I created an issue at https://github.com/scroll-tech/ceno/issues/751

Btw, we do have read only sections in the ELF already, and we should honour them.

**Review by matthiasgoergens — CHANGES_REQUESTED** [review]
Sorry, I prematurely removed my hold.  Please just use a `BTreeSet` or at least a `///` doc comment.

File: ceno_zkvm/src/e2e.rs (hunk 8 above)
Comment (matthiasgoergens) [reply ceno_zkvm/src/e2e.rs:192]: How do you know it's not related, if you don't understand it?

File: ceno_emul/src/platform.rs (hunk 3 above)
Comment (matthiasgoergens) [reply ceno_emul/src/platform.rs:14]: Interesting.  Thanks!

Then a `HashSet` is as good as the `BTreeSet`, but please use a doc comment for the explanation, so it shows up in our tooling.  Thanks!

File: ceno_emul/src/platform.rs [hunk 11]
```diff
@@ -10,20 +10,23 @@ use crate::addr::{Addr, RegIdx};
 #[derive(Clone, Debug)]
 pub struct Platform {
     pub rom: Range<Addr>,
-    pub ram: Range<Addr>,
+    // This is an `Option` to allow `const` here.
```
Comment (matthiasgoergens) [inline ceno_emul/src/platform.rs:13]:
```suggestion
    /// This is an `Option` to allow `pub const CENO_PLATFORM: Platform`
```

**Comment by matthiasgoergens** [comment]
I enabled auto-merge again.  (It got disabled, when I mucked around with my review.)

**Comment by matthiasgoergens** [comment]
@naure Sorry, the renaming of ROM to program code (or similar) is fine.  And I recognise that the changes you made to the `can_write` function is also fine in the sense that they make the emulator reflect what the circuits currently enforce.  (A promised in the PR description.)

My request to respect read-only memory is more an aspiration for the future development.

**Review by matthiasgoergens — APPROVED** [review]

**Pull request merged**